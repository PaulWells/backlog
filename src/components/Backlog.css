.title {
  margin-bottom: 10px;
}

.backlog {
  margin-top: 50px;
  text-align: center;
}

.logItemInput {
  margin-top: 10px;
  margin-botton: 10px;
}
