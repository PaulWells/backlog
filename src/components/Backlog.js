import React, { Component } from 'react';
import getStore from '../index.js';

var store;
getStore(function (createdStore) {
  store = createdStore;
});

class Backlog extends Component {

  render() {
    return (
      <div>
      <LogTitle/>
      {this.props.backlogState.listItems.map(function(item, i){
        return <LogItem text={item.text} index={i} completed={item.completed} key={i}/>;
      })}
      <LogItemInput value={this.props.backlogState.inputText}/>
      </div>
    );
  }
}

class LogTitle extends Component {
  render() {
    return (
      <div>
      Backlog
      </div>
    );
  }
}

class LogItemInput extends Component {

  constructor (props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
  }

  onKeyDown (event) {
    if (event.key === 'Enter') {
      store.dispatch({type: 'ADD_LIST_ITEM', text:this.props.value});
    }
  }

  handleChange (event) {
    store.dispatch({type: 'UPDATE_NEW_LIST_ITEM_INPUT', value: event.target.value})
  }

  render() {
    return (
      <div>
      <input type="text" value={this.props.value} onChange={this.handleChange} onKeyDown={this.onKeyDown}></input>
      </div>
    );
  }
}

class LogItem extends Component {

  render() {
    return (
      <div>
      <CheckBox checked={this.props.completed} index={this.props.index}/>
      <span style={{
              textDecoration:
              this.props.completed ?
                'line-through' :
                'none'
            }}>
            {this.props.text}
      </span>
      <DeleteIcon onDelete={() => store.dispatch({type: 'DELETE_LIST_ITEM', index: this.props.index })}/>
      </div>
    );
  }
}

class CheckBox extends Component {

  constructor() {
    super();
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange() {
    store.dispatch({type: 'TOGGLE_LIST_ITEM_COMPLETED', completed: !this.props.checked, index: this.props.index});
  }

  render() {
    return (
      <input type="checkbox" checked={this.props.checked} onChange={this.handleChange}/>
    );
  }
}

class DeleteIcon extends Component {
  render() {
    return (
      <button onClick={this.props.onDelete}>X</button>
    )
  }
}

export { Backlog }
