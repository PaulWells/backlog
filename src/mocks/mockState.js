import { currentTime } from '../utils/time';

const dateAdded = currentTime();

export const createMockListItem = (item) => {
  const defaultObject = {
    completed: false,
    text: "",
    id: 0,
    dateAdded
  }

  return Object.assign({}, defaultObject, item);
};

export const createMockListItems = (items) => {
  return items.map(function (item, i) {
    return createMockListItem(item);
  })
}

export const createMockState = (listItems = [], inputText = "") => ({
  inputText,
  listItems
});
