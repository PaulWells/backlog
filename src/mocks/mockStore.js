// export empty store
import { createStore } from 'redux';
import { createMockState, createMockListItems, createMockListItem } from './mockState';
import { moreThanOneDayAgo, moreThanTwoWeeksAgo } from '../utils/time';

const mockReducer = (state, action) => {
  return state;
};

const item1 = {text: "hello", id: 0};
const item2 = {text: "it's me", id: 1};
const expiredCompletedItem = {text: "I'm from california", id: 2, completed: true, dateCompleted: moreThanOneDayAgo()}
const oldItem = {text: "dreaming", id: 2, completed: false, dateAdded: moreThanTwoWeeksAgo()}

export const mockEmptyStore = createStore(mockReducer, createMockState());
export const mockSingleItemStore = createStore(mockReducer, createMockState(createMockListItems([item1])));
export const mockMultipleItemStore = createStore(mockReducer, createMockState(createMockListItems([item1, item2])));
export const mockExpiredCompletedItemStore = createStore(mockReducer, createMockState(createMockListItems([expiredCompletedItem])));
export const mockOldItemStore = createStore(mockReducer, createMockState(createMockListItems([oldItem])));
