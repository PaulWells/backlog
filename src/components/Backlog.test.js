import React from 'react';
import ReactDOM from 'react-dom';
import Backlog from './Backlog';
import { mockEmptyStore, mockSingleItemStore, mockMultipleItemStore, mockExpiredCompletedItemStore, mockOldItemStore } from '../mocks/mockStore';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
  (
    <Provider store={ mockEmptyStore }>
      <Backlog/>
    </Provider>
  ), div);
});

it('renders no items', () => {
  const tree = renderer.create(
    (
      <Provider store={ mockEmptyStore }>
        <Backlog/>
      </Provider>
    )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a single item', () => {
  const tree = renderer.create(
    (
      <Provider store={ mockSingleItemStore }>
        <Backlog/>
      </Provider>
    )).toJSON();
  expect(tree).toMatchSnapshot();
});


test('It renders a list of items', () => {
  const tree = renderer.create(
    (
      <Provider store={ mockMultipleItemStore }>
        <Backlog/>
      </Provider>
    )).toJSON();
  expect(tree).toMatchSnapshot();
});

test('It does not render items that were completed more than 24 hours ago', () => {
  const tree = renderer.create(
    (
      <Provider store={ mockExpiredCompletedItemStore }>
        <Backlog/>
      </Provider>
    )).toJSON();
  expect(tree).toMatchSnapshot();
});

test('It does not render items that were created more than 2 weeks ago', () => {
  const tree = renderer.create(
    (
      <Provider store={ mockOldItemStore }>
        <Backlog/>
      </Provider>
    )).toJSON();
  expect(tree).toMatchSnapshot();
});
