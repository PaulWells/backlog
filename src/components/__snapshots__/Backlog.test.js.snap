exports[`test It does not render items that were completed more than 24 hours ago 1`] = `
<div
  className="backlog">
  <div
    className="backlogContent">
    <div
      className="backlogListContent" />
    <div
      className="logItemInputContainer">
      <textarea
        autoFocus={true}
        className="logItemInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="type a reminder..."
        style={
          Object {
            "color": "plum",
          }
        }
        type="text"
        value="" />
    </div>
  </div>
</div>
`;

exports[`test It does not render items that were created more than 2 weeks ago 1`] = `
<div
  className="backlog">
  <div
    className="backlogContent">
    <div
      className="backlogListContent" />
    <div
      className="logItemInputContainer">
      <textarea
        autoFocus={true}
        className="logItemInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="type a reminder..."
        style={
          Object {
            "color": "plum",
          }
        }
        type="text"
        value="" />
    </div>
  </div>
</div>
`;

exports[`test It renders a list of items 1`] = `
<div
  className="backlog">
  <div
    className="backlogContent">
    <div
      className="backlogListContent">
      <div
        className="logItem">
        <button
          className="deleteLogItemIcon"
          onClick={[Function]}>
          X
        </button>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox" />
        <span
          className="logItemTextContainer">
          <span
            className="logItemText"
            style={
              Object {
                "color": "plum",
                "textDecoration": "none",
              }
            }>
            it\'s me
          </span>
        </span>
      </div>
      <div
        className="logItem">
        <button
          className="deleteLogItemIcon"
          onClick={[Function]}>
          X
        </button>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox" />
        <span
          className="logItemTextContainer">
          <span
            className="logItemText"
            style={
              Object {
                "color": "plum",
                "textDecoration": "none",
              }
            }>
            hello
          </span>
        </span>
      </div>
    </div>
    <div
      className="logItemInputContainer">
      <textarea
        autoFocus={true}
        className="logItemInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="type a reminder..."
        style={
          Object {
            "color": "plum",
          }
        }
        type="text"
        value="" />
    </div>
  </div>
</div>
`;

exports[`test renders a single item 1`] = `
<div
  className="backlog">
  <div
    className="backlogContent">
    <div
      className="backlogListContent">
      <div
        className="logItem">
        <button
          className="deleteLogItemIcon"
          onClick={[Function]}>
          X
        </button>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox" />
        <span
          className="logItemTextContainer">
          <span
            className="logItemText"
            style={
              Object {
                "color": "plum",
                "textDecoration": "none",
              }
            }>
            hello
          </span>
        </span>
      </div>
    </div>
    <div
      className="logItemInputContainer">
      <textarea
        autoFocus={true}
        className="logItemInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="type a reminder..."
        style={
          Object {
            "color": "plum",
          }
        }
        type="text"
        value="" />
    </div>
  </div>
</div>
`;

exports[`test renders no items 1`] = `
<div
  className="backlog">
  <div
    className="backlogContent">
    <div
      className="backlogListContent" />
    <div
      className="logItemInputContainer">
      <textarea
        autoFocus={true}
        className="logItemInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="type a reminder..."
        style={
          Object {
            "color": "plum",
          }
        }
        type="text"
        value="" />
    </div>
  </div>
</div>
`;
